### PR TITLE
[SYCL-MLIR] Implement SYCL `GetOp` and `GetScalarOp` types

### PR DESCRIPTION
**Pull request by whitneywhtsang**
This PR implements the SYCL `GetOp` and `GetScalarOp` types.

SYCL runtime implementation:
```
template <typename T> class GetOp {
public:
  using DataT = T;
  DataT getValue(size_t) const { return (DataT)0; }
  DataT operator()(DataT, DataT) { return (DataT)0; }
};

template <typename T> class GetScalarOp {
public:
  using DataT = T;
  GetScalarOp(DataT Data) : m_Data(Data) {}
  DataT getValue(size_t) const { return m_Data; }

private:
  DataT m_Data;
};
```

clang generated example:
```
%"class.sycl::_V1::detail::GetOp" = type { i8 }
%"class.sycl::_V1::detail::GetScalarOp" = type { i32 }
```

Signed-off-by: Tsang, Whitney <whitney.tsang@intel.com>

### CODE REVIEW

File: polygeist/tools/cgeist/Lib/TypeUtils.cc [hunk 1]
```diff
@@ -187,6 +187,15 @@ mlir::Type getSYCLType(const clang::RecordType *RT,
       return mlir::sycl::GroupType::get(CGT.getModule()->getContext(), Dim,
                                         Body);
     }
+    if (CTS->getName() == "GetOp") {
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/TypeUtils.cc:190]: No {}

**Review by etiotto — APPROVED** [review]
LGTM

File: mlir-sycl/lib/Dialect/IR/SYCLOpsAlias.cpp [hunk 2]
```diff
@@ -67,6 +67,10 @@ SYCLOpAsmInterface::getAlias(mlir::Type Type, llvm::raw_ostream &OS) const {
     OS << "sycl_group_" << Group.getDimension();
     return AliasResult::FinalAlias;
   }
+  if (const auto GetScalarOp = Type.dyn_cast<mlir::sycl::GetScalarOpType>()) {
+    OS << "sycl_get_scalar_op_" << GetScalarOp.getDataType();
```
Comment (victor-eds) [inline mlir-sycl/lib/Dialect/IR/SYCLOpsAlias.cpp:71]:
```suggestion
    OS << "sycl_get_scalar_op_" << GetScalarOp.getDataType() << "_";
```

We don't want to have numeric characters at the end of alias ([see](https://github.com/intel/llvm/pull/7559#discussion_r1034475004)). Type could be, e.g., `i32`.

File: mlir-sycl/test/Conversion/SYCLToLLVM/sycl-types-to-llvm.mlir [hunk 3]
```diff
@@ -98,6 +100,12 @@ module {
   func.func @test_group(%arg0: !sycl_group_1) {
     return
   }
+  func.func @test_get_op(%arg0: !sycl.get_op) {
+    return
+  }
+  func.func @test_get_scalar_op(%arg0: !sycl.get_scalar_op<[i32], (i32)>) {
+    return
+  }
```
Comment (victor-eds) [inline mlir-sycl/test/Conversion/SYCLToLLVM/sycl-types-to-llvm.mlir:108]: Can we use the alias instead as in the rest of cases?

File: mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLOps.td [hunk 4]
```diff
@@ -207,6 +207,16 @@ def SYCL_GroupType : SYCL_Type<"Group", "group"> {
   let assemblyFormat = "`<` `[` $dimension `]` `,` `(` $body `)` `>`";
 }
 
+def SYCL_GetOpType : SYCL_Type<"GetOp", "get_op"> {
+  let assemblyFormat = "";
+}
```
Comment (victor-eds) [inline mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLOps.td:212]: I'd say we need a `dataType` parameter here holding the data type.

File: mlir-sycl/lib/Dialect/IR/SYCLOpsAlias.cpp (hunk 2 above)
Comment (whitneywhtsang) [reply mlir-sycl/lib/Dialect/IR/SYCLOpsAlias.cpp:71]: If you look at the test cases, actually the alias has `_` at the end.

File: mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLOps.td (hunk 4 above)
Comment (whitneywhtsang) [reply mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLOps.td:212]: `GetOp` holds no data type.

Comment (victor-eds) [reply mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLOps.td:212]: I was thinking on a scenario in which we wanted to have `getValue` or `operator()` as operations in the dialect and add verifiers for those. If you think that'll never be the case, we can just merge this.

Comment (whitneywhtsang) [reply mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLOps.td]: ic, it can be useful, added.

**Review by victor-eds — APPROVED** [review]

**Pull request merged**